Extension/src/Utility/Eventing/emitter.ts

```diff
@@ -0,0 +1,279 @@
+
+    /**
+   * Creates a named event function for triggering events in an {@link Emitter}
+   * @template TText - a text string that can be used to filter events during subscription
+   * @template TData - the expected input type of the event handler function.
+   * @template TOutput - the expected output type of of the event handler function.
+   *
+   * @returns the created event trigger function of type `(text: TText, input:TData) => Promise<TOutput|EventStatus>`
+   *
+   * @remarks When the trigger is invoked, the event handler can return one of the following:
+   *  - `<TOutput>` - an value of type `<TOutput>` - the event is not cancelled, and the emitter can use the data returned.
+   *  - {@link Continue} (or `undefined`) - the event is not cancelled, and execution should continue normally.
+   *  - {@link Cancelled} - the event is requested to be cancelled (there is no guarantee however.)
+   *
+   * @param eventName - the string name of the event
+   * @param options - options for the event trigger (default: `{now: false}`)
+   */
+    protected newEvent<TText extends string, TData, TOutput>(eventName: string, options?: EventOptions<TOutput>): (text: TText, data: TData) => Promise<TOutput>;
+    protected newEvent<TOutput>(eventName: string, options?: EventOptions & WithDefault<TOutput>): () => Promise<TOutput>;
+    protected newEvent<TData, TOutput>(eventName: string, options?: EventOptions & WithDefault<TOutput>): (input: TData) => Promise<TOutput>;
+    protected newEvent<TText extends string, TData, TOutput>(eventName: string, options?: EventOptions<TOutput> & WithDefault<TOutput>): (text: TText, data: TData) => Promise<TOutput>;
+    protected newEvent<TText extends string, TData, TOutput>(eventName: string, options?: EventOptions<TOutput> & WithDefault<TOutput>): (text: TText, data: TData) => Promise<EventStatus | TOutput> {
+        eventName = smash(eventName);
+        this.#knownEvents.add(eventName);
+        const descriptors = options?.descriptors ? new Descriptors(this.descriptors, options.descriptors) : this.descriptors;
+        // is it an immediate event?
+        if (options?.now) {
+            return async (input?: TText, data?: TData): Promise<TOutput | EventStatus> => {
+                switch (options?.once) {
+                    case false:
+                        // already triggered this one time event.
+                        return;
+
+                    case true:
+                        options.once = false;
+                }
+                // trigger the event
+                const result = await ((data !== undefined) ?
+                    emitNow<TOutput>(eventName, descriptors, this, input || '', data as any) : // text and data
+                    emitNow<TOutput>(eventName, descriptors, this, input || '', input as any)); // text or data (or neither)
+
+                return is.cancelled(result) ?         // was the event cancelled?

```

Comment:
What's up with the random spacing before `//`?

Reply:
reducing whitespace to 1